src/rust/catloop/mod.rs

```diff
@@ -564,38 +612,84 @@ impl CatloopLibOS {
         Err(Fail::new(libc::EINVAL, &cause))
     }
 
+    /// Harvests resources on close.
+    fn harvest_resources_on_close(&mut self, qd: QDesc) -> Result<(), Fail> {
+        // If this was an async_close(), harvest resources.
+        let mut qtable: RefMut<IoQueueTable<CatloopQueue>> = self.qtable.borrow_mut();
+        match qtable.get_mut(&qd) {
+            Some(queue) => {
+                // Rollback the port allocation.
+                if let Socket::Active(Some(addr)) | Socket::Passive(addr) = queue.get_socket() {
+                    if EphemeralPorts::is_private(addr.port()) {
+                        if self.ephemeral_ports.borrow_mut().free(addr.port()).is_err() {
+                            // We fail if and only if we attempted to free a port that was not allocated.
+                            // This is unexpected, but if it happens, issue a warning and keep going,
```

Comment:
If it can only fail in this situation, should the message state that instead of leaking port?

Reply:
Sure. I added an extra commit message there.